apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-metricsdk-plugin/src/test/java/co/elastic/apm/agent/otelmetricsdk/AbstractOtelMetricsTest.java

```diff
@@ -0,0 +1,624 @@
+            .put(booleanArrayKey("bool_array_attrib"), List.of(true, false))
+            .build();
+
+        counter.add(42, attribs);
+
+        resetReporterAndFlushMetrics();
+        assertThatMetricSets(reporter.getBytes())
+            .hasMetricsetCount(1)
+            .hasMetricsetWithLabelsSatisfying(Map.of(
+                    "otel_instrumentation_scope_name", "test",
+                    "string_attrib", "foo",
+                    "double_attrib", Double.MAX_VALUE,
+                    "long_attrib", Long.MAX_VALUE,
+                    "bool_attrib", false
+                ), metrics -> metrics
+                    .hasMetricsCount(1)
+                    .containsValueMetric("my_counter", 42)
+            );
+    }
+
+    @Test
+    public void testAttributeNameSanitization() {
+        Meter testMeter = createMeter("test");
+        LongCounter counter = testMeter.counterBuilder("my_counter").build();
+
+        Attributes attributes = Attributes.of(stringKey("please*sanitize.me\""), "dont*sanitize.me\"");
+        counter.add(42, attributes);
+
+        resetReporterAndFlushMetrics();
+        assertThatMetricSets(reporter.getBytes())
+            .hasMetricsetCount(1)
+            .hasMetricsetWithLabelsSatisfying(
+                "otel_instrumentation_scope_name", "test",
+                "please_sanitize_me_", "dont*sanitize.me\""
+                , metrics -> metrics
+                    .hasMetricsCount(1)
+                    .containsValueMetric("my_counter", 42)
+            );
+    }
+
+    @Test
+    public void testTimestampPresent() {
+        Meter testMeter = createMeter("test");
+        LongCounter counter = testMeter.counterBuilder("my_counter").build();
+
+        long timestampMicros = System.currentTimeMillis() * 1000L;
+        counter.add(42);
+
+        resetReporterAndFlushMetrics();
+        assertThatMetricSets(reporter.getBytes())
+            .hasMetricsetCount(1)
+            .first()
+            //check for a range of +- 1 second to be safe against CI slowness / clock inaccuracy
+            .hasTimestampInRange(timestampMicros - 1_000_000L, timestampMicros + 1_000_000L);
+    }
+
+    @Test
+    public void testMetricsCombinedInMetricsets() {
+        Meter testMeter = createMeter("test");
+        ObservableDoubleGauge gauge1 = testMeter
+            .gaugeBuilder("my_gauge")
+            .buildWithCallback((obs) -> {
+                obs.record(42);
+                obs.record(142, Attributes.of(stringKey("foo"), "bar"));
+            });
+        ObservableLongGauge gauge2 = testMeter
+            .gaugeBuilder("my_other_gauge").ofLongs()
+            .buildWithCallback((obs) -> {
+                obs.record(43);
+                obs.record(143, Attributes.of(stringKey("foo"), "bar"));
+            });
+
+        resetReporterAndFlushMetrics();
+        assertThatMetricSets(reporter.getBytes())
+            .hasMetricsetCount(2)
+            .hasMetricsetWithLabelsSatisfying(
```

Comment:
[minor] maybe for assertion readability I would split this in two distinct parts: `hasLabels(Map.of(....))` and `satisfies(metrics->metrics. [...])`.

Reply:
Unfortunately it is not possible to switch this easily into two operations:
`reporter.getBytes()` returns all reported metricsets, which should be two in this case.
We could do a `hasMetricsetSatisfying(metricset -> metricset.hasLabels(...).hasMetricValue(..)`, but I decided against it: What do you print if no metricset satisfies the provided condition? With my approach, we can distinguish based on the error message wheterh (a) no metricset with the given labels exist and (b) the metricset with the labels has other bad contents.